File: app/src/main/java/com/prey/json/parser/JSONParser.java
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2015 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.json.parser;

import java.util.ArrayList;
import java.util.List;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import android.content.Context;

import com.prey.PreyLogger;
import com.prey.net.PreyHttpResponse;
import com.prey.net.PreyRestHttpClient;

public class JSONParser {

    JSONObject jObj;

    boolean error;

    private final static String COMMAND = "\"command\"";
    private final static String TARGET  = "\"target\"";

    public JSONParser() {
    }

    public List<JSONObject> getJSONFromUrl(Context ctx, String uri) {
        PreyLogger.d("getJSONFromUrl:" + uri);
        String sb=null;
        String json=null;
        try{
            PreyHttpResponse response=PreyRestHttpClient.getInstance(ctx).get(uri,null);
            try{sb=response.getResponseAsString();}catch(Exception e){PreyLogger.e("Error:"+e.getMessage(),e);}
            if (sb!=null)
                json = sb.trim();
        }catch(Exception e){
            PreyLogger.e("Error, causa:" + e.getMessage(), e);
            return null;
        }
        if(sb!=null) {
            PreyLogger.d("_______cmd________");
            PreyLogger.d(sb);
        }
        //json = "[{\"command\":\"history\",\"target\":\"call\",\"options\":{}}]";
        //json = "[{\"command\":\"history\",\"target\":\"sms\",\"options\":{}}]";
        //json = "[{\"command\":\"history\",\"target\":\"contact\",\"options\":{}}]";
        //json = "[{\"command\":\"start\",\"target\":\"system_install\",\"options\":{}}]";
        //json = "[{\"command\":\"start\",\"target\":\"server\",\"options\":{}}]";
        //json = "[{\"command\":\"start\",\"target\":\"ring\",\"options\":{}}]";
        //json = "[{\"command\":\"start\",\"target\":\"video\",\"options\":{}}]";
        //json = "[{\"command\":\"get\",\"target\":\"picture\",\"options\":{}}]";
        //json = "[{\"command\":\"get\",\"target\":\"report\",\"options\":{\"include\":[\"picture\"]}}]";
        //json = "[{\"command\":\"get\",\"target\":\"report\",\"options\":{\"include\":[\"picture\",\"location\",\"screenshot\",\"access_points_list\"]}}]";
        //json = "[{\"command\":\"get\",\"target\":\"report\",\"options\":{\"include\":[\"screenshot\",\"picture\",\"location\"]}}]";
        //json = "[{\"command\":\"get\",\"target\":\"report\",\"options\":{\"include\":[\"screenshot\"]}}]";
        //json = "[{\"command\":\"get\",\"target\":\"report\",\"options\":{\"include\":[\"location\"]}}]";
        //json = "[{\"command\":\"start\",\"target\": \"geofencing\",\"options\": {\"origin\": \"-70.60713481,-36.42372147\",\"radius\": \"100\" }}]";
        //json = "[{\"command\":\"start\",\"target\": \"geofencing\",\"options\": {\"origin\": \"-70.7193117,-32.7521112\",\"radius\": \"100\" }}]";
        //json = "[{\"command\":\"start\",\"target\": \"geofencing\",\"options\": {\"id\":\"id1\",\"origin\":\"-70.60713481,-33.42372147\",\"radius\":\"100\",\"type:\":\"in",\"expire":"-1" }}]";
        //json = "[{\"command\":\"stop\",\"target\": \"geofencing\",\"options\": {\"id\":\"id1\"}}]";
        //json = "[{\"command\":\"start\",\"target\":\"alert\",\"options\":{\"message\":\"This device i.\"}},{\"command\":\"start\",\"target\":\"alarm\",\"options\":null}, {\"command\":\"get\",\"target\":\"report\",\"options\":{\"include\"[\"picture\",\"location\",\"screenshot\",\"access_points_list\"]}}]";
        //json = "[{\"command\":\"start\",\"target\":\"alert\",\"options\":{\"message\":\"This device i.\"}}, {\"command\":\"get\",\"target\":\"report\",\"options\":{\"delay\": \"25\",\"include\"[\"picture\",\"location\",\"screenshot\",\"access_points_list\"]}}]";
        //json = "[{\"command\":\"start\",\"target\":\"alert\",\"options\":{\"message\":\"This device i.\"}}
        //json = "[{\"command\": \"get\",\"target\": \"location\",\"options\": {}}]";
        //json = "[{\"command\":\"start\",\"target\":\"alert\",\"options\":{\"message\":\"This device i.\"}}]";
        //json = "[{\"command\":\"start\",\"target\":\"alarm\",\"options\":null}]";
        //json = "[{\"command\":\"start\",\"target\":\"alarm\",\"options\":null}]";
        //json = "[{\"command\":\"get\",\"target\":\"report\",\"options\":{\"interval\":\"2\"}}]";
        //json = "[{\"command\":\"get\",\"target\":\"report\",\"options\":{\"include\":[\"picture\",\"location\",\"screenshot\",\"access_points_list\"],\"interval\":\"10\"}}]";
        //json = "[{\"command\":\"get\",\"target\":\"report\",\"options\":{\"include\":[\"picture\",\"location\",\"access_points_list\"],\"interval\":\"10\"}}]";
        //json = "[{\"command\":\"start\",\"target\":\"camouflage\",\"options\":null}]";
        //json = "[{\"command\":\"stop\",\"target\":\"camouflage\",\"options\":{\"interval\":\"2\"}}}]";
        //json = "[{\"target\":\"alert\",\"command\":\"start\",\"options\":{\"alert_message\":\"This device is stolen property. Please contact testforkhq@gmail.com to arrange its safe return.\"}},{\"target\":\"lock\",\"command\":\"start\",\"options\":{\"unlock_pass\":\"oso\"}},{\"command\":\"get\",\"target\":\"location\"},{\"target\":\"network\",\"command\":\"start\"},{\"target\":\"geo\",\"command\":\"start\"}]";
        //json = "[{\"command\":\"start\",\"target\":\"contacts_backup\" }]";
        //json = "[{\"command\":\"start\",\"target\":\"contacts_restore\" }]";
        //json = "[{\"command\":\"start\",\"target\":\"browser\" }]";
        //json = "[{\"command\":\"get\",\"target\":\"report\",\"options\":{\"interval\":\"2\",\"exclude\":[\"picture\",false]}}]";
        //json = "[{\"command\":\"start\",\"target\": \"detach\",\"options\": {}}]";
        if ("[]".equals(json)) {
            return null;
        }
        if ("Invalid.".equals(json)) {
            return null;
        }
        return getJSONFromTxt(ctx, json);
    }

    /**
     * Parse a JSON string into a list of JSONObject instances.
     *
     * @param context The application context.
     * @param jsonString The JSON string to parse.
     * @return A list of JSONObject instances, or null if the input string is invalid.
     */
    public List<JSONObject> getJSONFromTxt(Context context, String jsonString) {
        // Check if the input string is null or empty
        if (jsonString == null || jsonString.isEmpty() || jsonString.equals("Invalid data received") || jsonString.equals("[null]")) {
            // If so, return null immediately
            return null;
        }
        // Create an empty list to store the parsed JSONObject instances
        List<JSONObject> jsonObjectList = new ArrayList<>();
        // Wrap the input string in a JSON object with a "prey" key
        String wrappedJsonString = "{\"prey\":" + jsonString + "}";
        try {
            // Create a JSONObject instance from the wrapped string
            JSONObject jsonObject = new JSONObject(wrappedJsonString);
            // Get the JSONArray instance associated with the "prey" key
            JSONArray jsonArray = jsonObject.getJSONArray("prey");
            // Iterate over the JSONArray and create a JSONObject instance for each element
            for (int i = 0; i < jsonArray.length(); i++) {
                // Get the current element as a string
                String jsonElement = jsonArray.get(i).toString();
                // Create a JSONObject instance from the string
                JSONObject commandObject = new JSONObject(jsonElement);
                // Add the JSONObject instance to the list
                jsonObjectList.add(commandObject);
            }
        } catch (Exception e) {
            // Log any errors that occur during parsing
            PreyLogger.e(String.format("Error parsing JSON:%s", e.getMessage()), e);
        }
        // Return the list of parsed JSONObject instances
        return jsonObjectList;
    }

    public List<JSONObject> getJSONFromTxt2(Context ctx, String json) {
        jObj = null;
        List<JSONObject> listaJson = new ArrayList<JSONObject>();
        List<String> listCommands = getListCommands(json);
        for (int i = 0; listCommands != null && i < listCommands.size(); i++) {
            String command = listCommands.get(i);
            try {
                jObj = new JSONObject(command);
                listaJson.add(jObj);
            } catch (JSONException e) {
                PreyLogger.e("JSON Parser, Error parsing data " + e.toString(), e);
            }
        }
        PreyLogger.d("json:" + json);
        // return JSON String
        return listaJson;
    }

    private List<String> getListCommands(String json) {
        if (json.indexOf("[{"+COMMAND)==0){
            return getListCommandsCmd(json);
        }else{
            return getListCommandsTarget(json);
        }
    }

    private List<String> getListCommandsTarget(String json) {
        json = json.replaceAll("nil", "{}");
        json = json.replaceAll("null", "{}");
        List<String> lista = new ArrayList<String>();
        int posicion = json.indexOf(TARGET);
        json = json.substring(posicion + 8);
        posicion = json.indexOf(TARGET);
        String command = "";
        while (posicion > 0) {
            command = json.substring(0, posicion);
            json = json.substring(posicion + 8);
            lista.add("{" + TARGET + cleanChar(command));
            posicion = json.indexOf("\"target\"");
        }
        lista.add("{" + TARGET + cleanChar(json));
        return lista;
    }

    private List<String> getListCommandsCmd(String json) {
        json = json.replaceAll("nil", "{}");
        json = json.replaceAll("null", "{}");
        List<String> lista = new ArrayList<String>();
        int posicion = json.indexOf(COMMAND);
        json = json.substring(posicion + 9);
        posicion = json.indexOf(COMMAND);
        String command = "";
        while (posicion > 0) {
            command = json.substring(0, posicion);
            json = json.substring(posicion + 9);
            lista.add("{" + COMMAND + cleanChar(command));
            posicion = json.indexOf("\"command\"");
        }
        lista.add("{" + COMMAND + cleanChar(json));
        return lista;
    }

    private String cleanChar(String json) {
        if (json != null) {
            json = json.trim();
            char c = json.charAt(json.length() - 1);
            while (c == '{' || c == ',' || c == ']') {
                json = json.substring(0, json.length() - 1);
                json = json.trim();
                c = json.charAt(json.length() - 1);
            }
        }
        return json;
    }
}

File: app/src/main/java/com/prey/json/parser/JSONParser.kt
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2025 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.json.parser

import com.prey.PreyLogger
import com.prey.net.PreyRestHttpClient

import org.json.JSONObject

/**
 * A utility class for parsing JSON data.
 */
class JSONParser {
    var jsonObject: JSONObject? = null
    var error: Boolean = false

    /**
     * Retrieves JSON data from a URL and returns a list of JSON objects.
     *
     * @param context The application context.
     * @param uri The URL to retrieve JSON data from.
     * @return A list of JSON objects, or null if an error occurred.
     */
    fun getJSONFromUrl(preyRestHttpClient: PreyRestHttpClient, uri: String): List<JSONObject>? {
        PreyLogger.d("getJSONFromUrl:$uri")
        var sb: String? = null
        var json: String? = null
        try {
            val params = HashMap<String, String?>()
            val response = preyRestHttpClient.get(uri, params)
            try {
                if (response != null) {
                    sb = response.getResponseAsString()
                }
            } catch (e: Exception) {
                PreyLogger.e("Error: ${e.message}", e)
            }
            if (sb != null) json = sb.trim { it <= ' ' }
        } catch (e: Exception) {
            PreyLogger.e("Error: ${e.message}", e)
            return null
        }
        if (sb != null) {
            PreyLogger.d("_______cmd________")
            PreyLogger.d(sb)
        }
        if ("[]" == json) {
            return null
        }
        if ("Invalid." == json) {
            return null
        }
        return getJSONFromTxt(json!!)
    }

    /**
     * Parses a JSON string and returns a list of JSON objects.
     *
     * @param context The application context.
     * @param jsonString The JSON string to parse.
     * @return A list of JSON objects, or null if an error occurred.
     */
    fun getJSONFromTxt(jsonString: String): List<JSONObject>? {
        if ("Invalid JSON".equals(jsonString)) return null
        if ("Invalid data received".equals(jsonString)) return null
        if ("[null]".equals(jsonString)) return null
        if ("".equals(jsonString)) return null
        val jsonList: MutableList<JSONObject> = ArrayList()
        PreyLogger.d("jsonString:${jsonString}")
        try {
            val jsonObject = JSONObject("{\"prey\":$jsonString}")
            val jsonArray = jsonObject.getJSONArray("prey")
            for (i in 0 until jsonArray.length()) {
                val jsonCommand = jsonArray[i].toString()
                val commandObject = JSONObject(jsonCommand)
                PreyLogger.d(commandObject.toString())
                jsonList.add(commandObject)
            }
        } catch (e: Exception) {
            PreyLogger.e("Error: ${e.message}", e)
        }
        return jsonList
    }

    /**
     * Retrieves a list of commands from a JSON string.
     *
     * @param json The JSON string to parse.
     * @return A list of commands, or an empty list if the input string is invalid.
     */
    fun getListCommands(json: String): List<String> {
        return if (json.indexOf("[{${COMMAND}") == 0) {
            extractCommandsFromCommandJson(json)
        } else {
            extractCommandsFromTargetJson(json)
        }
    }

    /**
     * Retrieves a list of commands from a JSON string that starts with a target array.
     *
     * @param json The JSON string to parse.
     * @return A list of commands.
     */
    fun extractCommandsFromTargetJson(json: String): List<String> {
        var jsonString = json.replace("nil".toRegex(), "{}")
        jsonString = jsonString.replace("null".toRegex(), "{}")
        val commands: MutableList<String> = ArrayList()
        var position = jsonString.indexOf(TARGET)
        jsonString = jsonString.substring(position + 8)
        position = jsonString.indexOf(TARGET)
        var command = ""
        while (position > 0) {
            command = jsonString.substring(0, position)
            jsonString = jsonString.substring(position + 8)
            commands.add("{${TARGET}${cleanChar(command)}")
            position = jsonString.indexOf("\"target\"")
        }
        commands.add("{${TARGET}${cleanChar(jsonString)}")
        return commands
    }

    /**
     * Retrieves a list of commands from a JSON string that starts with a command array.
     *
     * @param json The JSON string to parse.
     * @return A list of commands.
     */
    fun extractCommandsFromCommandJson(json: String): List<String> {
        var jsonString = json.replace("nil".toRegex(), "{}")
        jsonString = jsonString.replace("null".toRegex(), "{}")
        val jsonList: MutableList<String> = ArrayList()
        var position = jsonString.indexOf(COMMAND)
        jsonString = jsonString.substring(position + 9)
        position = jsonString.indexOf(COMMAND)
        var command = ""
        while (position > 0) {
            command = jsonString.substring(0, position)
            jsonString = jsonString.substring(position + 9)
            jsonList.add("{${COMMAND}${cleanChar(command)}")
            position = jsonString.indexOf("\"command\"")
        }
        jsonList.add("{${COMMAND}${cleanChar(jsonString)}")
        return jsonList
    }

    /**
     * Cleans a JSON string by removing trailing characters.
     *
     * @param json The JSON string to clean.
     * @return The cleaned JSON string.
     */
    fun cleanChar(json: String): String? {
        var jsonString: String? = json
        if (jsonString != null) {
            jsonString = jsonString.trim { it <= ' ' }
            var c = jsonString[jsonString.length - 1]
            while (c == '{' || c == ',' || c == ']') {
                jsonString = jsonString!!.substring(0, jsonString.length - 1)
                jsonString = jsonString.trim { it <= ' ' }
                c = jsonString[jsonString.length - 1]
            }
        }
        return jsonString
    }

    companion object {
        private const val COMMAND = "\"command\""
        private const val TARGET = "\"target\""
    }

}